### PR TITLE
add vertexai/featureonlinestore resource

### PR DESCRIPTION
**Pull request by kkram01**
This PR aims to add support to vertex AI FeatureOnlineStore resource which is recently launched. https://github.com/hashicorp/terraform-provider-google/issues/16507

```release-note:enhancement
vertexai: added `force_destroy` field to `google_vertex_ai_feature_online_store` resource
```
```release-note:enhancement
vertexai: added `optimized`, `dedicated_serving_endpoint` and `embedding_management` fields to `google_vertex_ai_feature_online_store` resource (beta)
```
```release-note:bug
vertexai: changed `region` field from required to optional in `google_vertex_ai_feature_online_store` resource
```

### CODE REVIEW

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_optimized.tf.erb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  dedicated_serving_endpoint {
+    private_service_connect_config {
+       enable_private_service_connect = true
+       project_allowlist = [data.google_project.project.number]
+    }
+ }
+  force_destroy = true
+}
+
+data "google_project" "project" {}
```
Comment (SarahFrench) [inline mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_optimized.tf.erb]:
```suggestion
data "google_project" "project" {
  provider = "google-beta"
}
```

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ dedicated_serving_endpoint {
+   private_service_connect_config {
+      enable_private_service_connect = true
+      project_allowlist = [data.google_project.project.number]
+   }
+}
+  force_destroy = true
+}
+
+data "google_project" "project" {}
```
Comment (SarahFrench) [inline mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb]:
```suggestion
data "google_project" "project" {
  provider = "google-beta"
}```

**Review by SarahFrench — COMMENTED** [review]
I'm at the end of my day so can review properly tomorrow, but wanted to share how to fix the test failure affecting the Beta-only field tests:

```
    vcr_utils.go:152: Step 1/2 error: Error running pre-apply refresh: 
        Error: Inconsistent dependency lock file
        
        The following dependency selections recorded in the lock file are
        inconsistent with the current configuration:
          - provider registry.terraform.io/hashicorp/google: required by this configuration but no version is selected
        
        To make the initial dependency selections that will initialize the dependency
        lock file, run:
          terraform init
```

 When testing Beta features only the `google-beta` provider is available for use, so all resources etc need to use it. The data sources are defaulting to looking for the GA `google` provider but the test framework doesn't make it available.

**Comment by SarahFrench** [comment]
We had good timing then! 😄

File: mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+			},
+		},
+	})
+}
+
+func testAccVertexAIFeatureOnlineStore_vertexAiFeatureonlinestoreWithBigtable_basic(context map[string]interface{}) string {
+	return acctest.Nprintf(`
+##FeatureOnlineStore With BigTable
+resource "google_vertex_ai_feature_online_store" "featureonlinestore_bigtable" {
+  name     = "terraform2%{random_suffix}"
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go:52]: Could you please update this to have the prefix `tf-test-`? This helps us to automatically identify and delete resources that are created by acceptance tests but not deleted correctly when the test finishes.

File: mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  force_destroy = true
+}
+`, context)
+}
+
+func testAccVertexAIFeatureOnlineStore_vertexAiFeatureonlinestoreWithBigtableExample_update(context map[string]interface{}) string {
+	return acctest.Nprintf(`
+##FeatureOnlineStore With BigTable
+resource "google_vertex_ai_feature_online_store" "featureonlinestore_bigtable" {
+  name     = "terraform2%{random_suffix}"
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go:73]: See other comment

**Review by SarahFrench — COMMENTED** [review]
The [error shown in this message](https://github.com/GoogleCloudPlatform/magic-modules/pull/9632#issuecomment-1855244425) is :

```
  | After applying this test step, the plan was not empty.
  | stdout:
  | 
  | 
  | Terraform used the selected providers to generate the following execution
  | plan. Resource actions are indicated with the following symbols:
  |   ~ update in-place
  | 
  | Terraform will perform the following actions:
  | 
  |   # google_vertex_ai_feature_online_store.featureonlinestore will be updated in-place
  |   ~ resource "google_vertex_ai_feature_online_store" "featureonlinestore" {
  |         id               = "projects/PROJECT_ID/locations/us-central1/featureOnlineStores/terraform36zdh3a98t0"
  |         name             = "terraform36zdh3a98t0"
  |         # (8 unchanged attributes hidden)
  | 
  |       ~ dedicated_serving_endpoint {
  |             # (1 unchanged attribute hidden)
  | 
  |           + private_service_connect_config {
  |               + enable_private_service_connect = true
  |               + project_allowlist              = [
  |                   + "1234567890",
  |                 ]
  |             }
  |         }
  | 
  |         # (2 unchanged blocks hidden)
  |     }
  | 
  | Plan: 0 to add, 1 to change, 0 to destroy.
```

This is because, despite the POST request to create that resource including:

```
 "dedicatedServingEndpoint": {
  "privateServiceConnectConfig": {
   "enablePrivateServiceConnect": true,
   "projectAllowlist": [
    "1234567890"
   ]
  }
 },
```

the GET request to read the resource's information (after the create operation reported that it was done) includes this instead, and information from above is missing:

```
  "dedicatedServingEndpoint": {
    "publicEndpointDomainName": "523084585.us-central1-1051995038222.vdb.vertexai.goog"
  },
```

Do you know if this is expected behaviour from the Beta API?

File: mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go (hunk 3 above)
Comment (kkram01) [reply mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go]: Done

**Comment by SarahFrench** [comment]
Following the last error about `"description": "feature_online_store_id must begin with a lowercase or underscore and can only contain lowercase, numbers, and underscores.",
`, the prefix I requested can also be the form `tf_test_`

**Comment by kkram01** [comment]
@SarahFrench can you please check the above test fail of `TestAccComputeRegionPerInstanceConfig_removeInstanceOnDestroy`. It is not a part of the feature or code changes I was working on. TIA

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        immutable: true
+  - !ruby/object:Api::Type::NestedObject
+    name: 'bigtable'
+    description: |
+      Contains settings for the Cloud Bigtable instance that will be created to serve featureValues for all FeatureViews under this FeatureOnlineStore.
+    conflicts:
+      - optimized
+    exactly_one_of:
+      - bigtable
+      - optimized
```
Comment (SarahFrench) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:189]: The properties `conflicts` and `exactly_one_of` are very similar so we can remove one of them here (and in the description of the optimized field). When deciding which one to remove we need to decide whether it's valid for users to have both bigtable and optimized unset in their configuration.

When I created the resource with neither field set I got this error:

```
  "error": {
    "code": 400,
    "message": "FeatureOnlineStore must have either Bigtable or Optimized storage enabled.",
    "status": "INVALID_ARGUMENT"
  }
```

So I think we should keep `exactly_one_of` and remove `conflicts` from here

**Comment by SarahFrench** [comment]
`TestAccComputeRegionPerInstanceConfig_removeInstanceOnDestroy` failing above isn't related to this PR, so we can safely ignore it.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 5 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml]: Done

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+              The minimum number of nodes to scale down to. Must be greater than or equal to 1.
+          - !ruby/object:Api::Type::Integer
+            name: 'maxNodeCount'
+            required: true
+            description: |
+              The maximum number of nodes to scale up to. Must be greater than minNodeCount, and less than or equal to 10 times of 'minNodeCount'.
+          - !ruby/object:Api::Type::Integer
+            name: 'cpuUtilizationTarget'
+            description: |
+              A percentage of the cluster's CPU capacity. Can be from 10% to 80%. When a cluster's CPU utilization exceeds the target that you have set, Bigtable immediately adds nodes to the cluster. When CPU utilization is substantially lower than the target, Bigtable removes nodes. If not set will default to 50%.
```
[the remark this answers is not in the document]
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml:204]: Thanks for pointing. Updated!

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+      - !ruby/object:Api::Type::NestedObject
+        name: 'privateServiceConnectConfig'
+        description: |
+          Private service connect config.
+        default_from_api: true
+        properties:
+          - !ruby/object:Api::Type::Boolean
+            name: 'enablePrivateServiceConnect'
+            required: true
+            default_from_api: true
```
[the remark this answers is not in the document]
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml]: right! removed `default_from_api` here.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+            description: |
+              If set to true, customers will use private service connection to send request. Otherwise, the connection will set to public endpoint.
+          - !ruby/object:Api::Type::Array
+            name: 'projectAllowlist'
+            default_from_api: true
```
[the remark this answers is not in the document]
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml]: Removed the field here as well. I thought the default here would be an empty array from the API if not provided because the field is not required but I do now think after running some API calls that there is no API response for that field.

**Comment by SarahFrench** [comment]
@kkram01 I just found PR https://github.com/GoogleCloudPlatform/magic-modules/pull/9519 is working to implement the `google_vertexai_feature_online_store` resource. I'm going to talk to the reviewer of that PR about what to do. Given that PR #9519 is ~2 weeks older than _this_ PR I'm going to suggest that #9519 implements the resource and this PR will add the Beta-only fields.

**Comment by SarahFrench** [comment]
Follow up to my last message, do you work on a service team in Google that's responsible for this resource in Google Cloud?

**Comment by kkram01** [comment]
> Follow up to my last message, do you work on a service team in Google that's responsible for this resource in Google Cloud?

Hi, Thanks for pointing about the other PR. Yes, I'm working with our product team of vertex AI to get these created on priority. I'm okay with getting the GA resource created from the other PR too. I'd like to request to get featureonlinestore merged on priority because feature view resource is dependent on this to be developed. Thanks!

**Comment by SarahFrench** [comment]
@kkram01 I've assigned @/roaks3 as reviewer here, so he can coordinate the two PRs.

**Comment by roaks3** [comment]
@kkram01 I've just merged https://github.com/GoogleCloudPlatform/magic-modules/pull/9519, so whenever you have a chance, we should rebase this PR on top of those changes, which will likely require some manual work. Once the PR is back in a good state, I'll take another review pass here.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 9]
```diff
@@ -15,11 +15,13 @@
 name: FeatureOnlineStore
 base_url: projects/{{project}}/locations/{{region}}/featureOnlineStores
 create_url: projects/{{project}}/locations/{{region}}/featureOnlineStores?featureOnlineStoreId={{name}}
-self_link: projects/{{project}}/locations/{{region}}/featureOnlineStores/{{name}}
+self_link: 'projects/{{project}}/locations/{{region}}/featureOnlineStores/{{name}}'
```
Comment (roaks3) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:18]: Nit: Could be removed from the diff

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 10]
```diff
@@ -41,65 +43,183 @@ description: Vertex AI Feature Online Store provides a centralized repository fo
 autogen_async: false
 examples:
   - !ruby/object:Provider::Terraform::Examples
-    name: vertex_ai_feature_online_store
-    primary_resource_id: feature_online_store
+    name: 'vertex_ai_featureonlinestore_with_bigtable'
+    primary_resource_id: 'featureonlinestore'
+    primary_resource_name: "fmt.Sprintf(\"terraform%s\",
```
Comment (roaks3) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:48]: I don't think you need this field unless there are IAM resources being generated. Same for other examples

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 11]
```diff
@@ -41,65 +43,183 @@ description: Vertex AI Feature Online Store provides a centralized repository fo
...
+      context[\"random_suffix\"\
+      ])"
     vars:
-      name: example_feature_online_store
+      name: 'terraform2'
```
Comment (roaks3) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:52]: These variables need to have a `-` or `_` in order for them to behave properly in tests (it signals that a `tf-test` prefix and a random suffix should be added). I would suggest sticking with the prior name here because it is more descriptive, but you could change to something else as well. Same for other variables that are identifiers.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 12]
```diff
@@ -41,65 +43,183 @@ description: Vertex AI Feature Online Store provides a centralized repository fo
...
+      project: 'appeng-flex'
```
Comment (roaks3) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:53]: I don't see a project being used in this example, or any of the examples actually. Do you need it here? Same question for both `test_env_vars`: `org_id` and `billing_account`.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 13]
```diff
@@ -41,65 +43,183 @@ description: Vertex AI Feature Online Store provides a centralized repository fo
...
+      'If set to true, any FeatureViews and Features for this FeatureOnlineStore will also be deleted.'
+    default_value: false
 parameters:
   - !ruby/object:Api::Type::String
     name: region
     description: The region of feature online store. eg us-central1
     url_param_only: true
     immutable: true
-    required: true
+    default_from_api: true
```
Comment (roaks3) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:103]: Could you explain this change? This seems incorrect based on the url structure.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 14]
```diff
@@ -41,65 +43,183 @@ description: Vertex AI Feature Online Store provides a centralized repository fo
...
 properties:
   - !ruby/object:Api::Type::String
-    name: name
-    description: The resource name of the Feature Online Store.
+    name: 'name'
```
Comment (roaks3) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:106]: Nit: `'`s could be removed from the diff

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 15]
```diff
@@ -41,65 +43,183 @@ description: Vertex AI Feature Online Store provides a centralized repository fo
...
+    description:
+      The name of the featureOnlinestore. This value may be up to 60 characters, and
+      valid characters are [a-z0-9_]. The first character cannot be a number.
     immutable: true
-    required: true
+    url_param_only: true
```
Comment (roaks3) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:111]: Do you know if the configuration for `name` needs to change as you've suggested here? My understanding is that the previous config was working because the `name` can be provided in the request body, and it is returned in the response body, so it is arguably a simpler approach.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 16]
```diff
@@ -41,65 +43,183 @@ description: Vertex AI Feature Online Store provides a centralized repository fo
...
+    pattern: projects/{{project}}/locations/{{region}}/featureOnlineStores/{{name}}
     custom_flatten: templates/terraform/custom_flatten/name_from_self_link.erb
   - !ruby/object:Api::Type::String
-    name: createTime
```
Comment (roaks3) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:63]: Could you please re-order these fields to match the existing resource, to minimize no-op changes in the diff. The `bigtable` field changes below are particularly hard to follow.

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb [hunk 17]
```diff
@@ -0,0 +1,23 @@
+resource "google_vertex_ai_feature_online_store" "<%= ctx[:primary_resource_id] %>" {
+  provider = google-beta
+  name     = "<%= ctx[:vars]['name'] %>"
+  labels = {
+    foo = "bar"
+  }
+  region   = "us-central1"
+  bigtable {
+   auto_scaling {
```
Comment (roaks3) [inline mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb:9]: Nit: inconsistent indentation. Note that these examples will be visible at the top of the documentation website for this resource.

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_bigtable.tf.erb [hunk 18]
```diff
@@ -0,0 +1,16 @@
+##FeatureOnlineStore With BigTable
```
Comment (roaks3) [inline mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_bigtable.tf.erb:1]: Comment should be removed

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_bigtable.tf.erb [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+  }
+  region   = "us-central1"
+  bigtable {
+    auto_scaling {
+        min_node_count = 1
+        max_node_count = 3
+        cpu_utilization_target = 50
+    }
+  }
+  force_destroy = true
```
Comment (roaks3) [inline mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_bigtable.tf.erb:15]: This will show up on our website as an example, so I just wanted to make sure, do we want to instruct users to set `force_destroy = true` by default? Since it is false by default, my assumption for a field like this would be for users to use caution when setting it true.

File: mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go [hunk 20]
```diff
@@ -6,78 +6,83 @@ import (
...
+	"github.com/hashicorp/terraform-provider-google/google/envvar"
 )
 
-func TestAccVertexAIFeatureOnlineStore_updated(t *testing.T) {
+func TestAccVertexAIFeatureOnlineStore_vertexAiFeatureonlinestoreWithBigtable_updated(t *testing.T) {
 	t.Parallel()
 
 	context := map[string]interface{}{
-		"random_suffix": acctest.RandString(t, 10),
+		"org_id":          envvar.GetTestOrgFromEnv(t),
```
Comment (roaks3) [inline mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go:16]: Similar to the example comments, I don't see `org_id` or `billing_account` being used. Do you need them?

File: mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go [hunk 21]
```diff
@@ -6,78 +6,83 @@ import (
...
-        label-one = "value-one"
-    }
-
-    bigtable {
-        auto_scaling {
-            min_node_count = 1
-            max_node_count = 2
-            cpu_utilization_target = 60
-        }
+##FeatureOnlineStore With BigTable
```
Comment (roaks3) [inline mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go:50]: Comment should be removed

**Review by roaks3 — CHANGES_REQUESTED** [review]
Some initial comments that are largely cleanup or related to the rebase. Once those are resolved, I can take another pass with more focus on the specific field additions, but I expect those are probably in a good state based on previous review and passing tests.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 9 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml]: Done

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 10 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml]: Thanks! I see "primary_resource_name" is not being used in my examples. Updated!

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 12 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml]: Ah, It was a miss as I assumed a test project would be provisioned to run the changes or similar, since this was present in other resources I've seen where they are not being utilized. This is now updated.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 13 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml:82]: I've updated it to use default_from_api : true because if region isn't provided, I wanted to use the default region set on the project.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 14 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml]: Done

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 15 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml:89]: I've done this so that it is consistent with the legacy FeatureStore which has similar url structure and json request object. If the existing one is simpler and easy to understand, I can change it too. Pls let me know.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 16 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml:63]: Done!

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb (hunk 17 above)
Comment (kkram01) [reply mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb]: Updated the identation of all the examples with a terraform fmt.

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_bigtable.tf.erb (hunk 19 above)
Comment (kkram01) [reply mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_bigtable.tf.erb:15]: Yeah, You're right. The intention was to show that this is a field that could be set in one of the examples, but since it would be in the documentation anyway, I've removed it now.

File: mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go (hunk 20 above)
Comment (kkram01) [reply mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go]: Updated!

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 13 above)
Comment (roaks3) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml:82]: Ah right, thank you. This was incorrect in the previous PR then, but luckily it is not a breaking change.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 15 above)
Comment (roaks3) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml:89]: The precedence for vertexai is a bit fragmented unfortunately, with Endpoint being the primary exception to the pattern you mention. However, since most of the other resources do use `url_param_only` and `pattern`, I think it makes sense to keep them here.

However, are you sure we should remove `required`? What happens if a name is not specified?

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 22]
```diff
@@ -103,3 +133,64 @@ properties:
               A percentage of the cluster's CPU capacity. Can be from 10% to 80%.
               When a cluster's CPU utilization exceeds the target that you have set,
               Bigtable immediately adds nodes to the cluster. When CPU utilization is substantially lower than the target, Bigtable removes nodes. If not set will default to 50%.
+  - !ruby/object:Api::Type::NestedObject
+    name: optimized
+    description:
+      Contains settings for the Optimized store that will be created to serve featureValues for all FeatureViews under this FeatureOnlineStore
+    conflicts:
+      - embeddingManagement
+    exactly_one_of:
```
Comment (roaks3) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:142]: I believe you will need to update `bigtable` to include this same `exactly_one_of` block.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml [hunk 23]
```diff
@@ -103,3 +133,64 @@ properties:
...
+    name: dedicatedServingEndpoint
+    description: |
+       The dedicated serving endpoint for this FeatureOnlineStore, which is different from common vertex service endpoint. Only need to set when you choose Optimized storage type or enable EmbeddingManagement. Will use public endpoint by default.
+    default_from_api: true
+    min_version: beta
+    properties:
+      - !ruby/object:Api::Type::String
+        name: publicEndpointDomainName
+        description: |
+          populated with the domain name to use for this FeatureOnlineStore
```
Comment (roaks3) [inline mmv1/products/vertexai/FeatureOnlineStore.yaml:162]: This will also show up in the docs. These should generally be sentences, with at least the first character capitalized. I would personally suggest removing "populated with", since it isn't commonly included in these descriptions, and it is redundant with the "(Output)" prefix that will be generated.

Similar suggestion for some other fields below.

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+    auto_scaling {
+      min_node_count         = 1
+      max_node_count         = 2
+      cpu_utilization_target = 80
+    }
+  }
+  embedding_management {
+    enabled = true
+  }
+  force_destroy = true
```
Comment (roaks3) [inline mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb:18]: Should probably be removed, this will show up in our docs.

Follow-up question to confirm: when this config is run in our tests, I assume it will still be deleted without `force_destroy` because it is not using any feature views or features. Is that correct, or will there be leftover resources?

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_bigtable.tf.erb (hunk 18 above)
Comment (roaks3) [inline mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_bigtable.tf.erb:1]: I think this file should be removed now?

File: mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go [hunk 25]
```diff
@@ -18,7 +18,6 @@ func TestAccVertexAIFeatureOnlineStore_updated(t *testing.T) {
 	acctest.VcrTest(t, resource.TestCase{
 		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
 		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
-		CheckDestroy:             testAccCheckVertexAIFeatureOnlineStoreDestroyProducer(t),
```
Comment (roaks3) [inline mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go:21]: Do you know why this is needed? I'm wondering if this means resources will accumulate in our test environment. If so, we probably need to alter the test.

**Review by roaks3 — CHANGES_REQUESTED** [review]
A couple more small suggestions, but the main remaining question I have is how these tests will behave in our test environment. Namely, do we need to do anything special to keep these resources from accumulating, or will the test properly clean them up?

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb (hunk 24 above)
Comment (kkram01) [reply mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb:18]: > Follow

1. Can we have force_destroy = true as set for one of the examples ? This would also mean that we'd be addressing the missing test report that is getting generated for `force_destroy` field and also an example of the usage of the field too. 
2. Yes, It'll still be deleted. I've run the tests manually in my test env and the resources got deleted as expected at the end of the test in both force_destroy = true and false cases.

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 22 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml:146]: Yeah, thanks for pointing it out. I had it originally but seems to have gotten removed during rebase. Added it back

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 23 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml]: Updated!

File: mmv1/products/vertexai/FeatureOnlineStore.yaml (hunk 15 above)
Comment (kkram01) [reply mmv1/products/vertexai/FeatureOnlineStore.yaml:90]: Correct! The REST API documentation doesn't explicitly mention it as required field but when I try to run without this field specified, it throws an error. So, updated to include `required : true` now

File: mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go (hunk 25 above)
Comment (kkram01) [reply mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go:21]: I suppose this function verifies if the resources are actually deleted after the terraform destroy is complete. But Our update test documentation mentions to remove this when creating the update test case from generated test files (https://googlecloudplatform.github.io/magic-modules/develop/test/#add-an-update-test). Hence, excluded.

File: mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb (hunk 24 above)
Comment (roaks3) [reply mmv1/templates/terraform/examples/vertex_ai_featureonlinestore_with_beta_fields_bigtable.tf.erb:18]: Sure, if you feel it is better to include for this example I think that makes sense, and can always be adjusted later. Thanks for checking that resources get deleted!

File: mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go (hunk 25 above)
Comment (roaks3) [reply mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go:21]: Ah, that could probably be a bit more clear. Those instructions are saying to delete the function definition, because it is already being autogenerated elsewhere in the package, but you want to keep the call here for the reason you mentioned (it makes sure the resource was deleted).

Comment (kkram01) [reply mmv1/third_party/terraform/services/vertexai/resource_vertex_ai_feature_online_store_test.go]: Oh I see. Understood! Updated.

**Review by roaks3 — APPROVED** [review]
LGTM pending test results

**Comment by roaks3** [comment]
Note: updated the release notes
